Mathlib/MeasureTheory/Integral/SobolevInequality.lean

```diff
@@ -0,0 +1,610 @@
+* `MeasureTheory.snorm_le_snorm_fderiv_of_le`:
+  The bound holds when `1 ≤ p < n`, `0 ≤ q` and `p⁻¹ - n⁻¹ ≤ q⁻¹`.
+  Note that in this case the constant depends on the support of `u`.
+-/
+
+open scoped Classical BigOperators ENNReal NNReal Topology
+open Set Function Finset MeasureTheory Measure Filter
+
+noncomputable section
+
+variable {ι : Type*} [Fintype ι] [DecidableEq ι]
+
+local prefix:max "#" => Fintype.card
+
+/-! ## The grid-lines lemma -/
+
+variable {A : ι → Type*} [∀ i, MeasurableSpace (A i)]
+  (μ : ∀ i, Measure (A i)) [∀ i, SigmaFinite (μ i)]
+
+namespace MeasureTheory
+
+namespace GridLines
+
+/-- The "grid-lines operation" (not a standard name) which is central in the inductive proof of the
+Sobolev inequality.
+
+For a finite dependent product `Π i : ι, A i` of sigma-finite measure spaces, a finite set `s` of
+indices from `ι`, and a (later assumed nonnegative) real number `p`, this operation acts on a
+function `f` from `Π i, A i` into the extended nonnegative reals.  The operation is to partially
+integrate, in the `s` co-ordinates, the function whose value at `x : Π i, A i` is obtained by
+multiplying a certain power of `f` with the product, for each co-ordinate `i` in `s`, of a certain
+power of the integral of `f` along the "grid line" in the `i` direction through `x`.
+
+We are most interested in this operation when the set `s` is the universe in `ι`, but as a proxy for
+"induction on dimension" we define it for a general set `s` of co-ordinates: the `s`-grid-lines
+operation on a function `f` which is constant along the co-ordinates in `sᶜ` is morally (that is, up
+to type-theoretic nonsense) the same thing as the universe-grid-lines operation on the associated
+function on the "lower-dimensional" space `Π i : s, A i`. -/
+def T (p : ℝ) (f : (∀ i, A i) → ℝ≥0∞) (s : Finset ι) : (∀ i, A i) → ℝ≥0∞ :=
+  ∫⋯∫⁻_s, f ^ (1 - (s.card - 1 : ℝ) * p) * ∏ i in s, (∫⋯∫⁻_{i}, f ∂μ) ^ p ∂μ
+
+variable {p : ℝ}
+
+@[simp] lemma T_univ (f : (∀ i, A i) → ℝ≥0∞) (x : ∀ i, A i) :
+    T μ p f univ x =
+    ∫⁻ (x : ∀ i, A i), (f x ^ (1 - (#ι - 1 : ℝ) * p)
+    * ∏ i : ι, (∫⁻ t : A i, f (update x i t) ∂(μ i)) ^ p) ∂(.pi μ) := by
+  simp [T, lmarginal_univ, lmarginal_singleton, card_univ]
+
+@[simp] lemma T_empty (f : (∀ i, A i) → ℝ≥0∞) (x : ∀ i, A i) :
+    T μ p f ∅ x = f x ^ (1 + p) := by
+  simp [T]
+
+/-- The main inductive step in the grid-lines lemma for the Gagliardo-Nirenberg-Sobolev inequality.
+
+The grid-lines operation `GridLines.T` on a nonnegative function on a finitary product type is
+less than or equal to the grid-lines operation of its partial integral in one co-ordinate
+(the latter intuitively considered as a function on a space "one dimension down"). -/
+theorem T_insert_le_T_lmarginal_singleton (hp₀ : 0 ≤ p) (s : Finset ι)
+    (hp : (s.card : ℝ) * p ≤ 1)
+    (i : ι) (hi : i ∉ s) {f : (∀ i, A i) → ℝ≥0∞} (hf : Measurable f) :
+    T μ p f (insert i s) ≤ T μ p (∫⋯∫⁻_{i}, f ∂μ) s := by
+  calc T μ p f (insert i s)
```

Comment:
Whenever a proof is more than 15 lines long, I think it deserves some comments explaining what you are doing. This one qualifies.